### PR TITLE
Enforce unique index in bookmark table

### LINKED ISSUE #2218: Unique constraint broken in SQL causes crash
## Issue explanation (write below this line)

accidently tried to add an existing bookmark a second time. this is a 
reproducable crash on my device.

## Exception
* __User Action:__ UI Error
* __Request:__ Application crash
* __Version:__ 3.5.3
* __OS:__ Linux Android 10 - 29
* __Device:__ gta4xlwifi
* __Model:__ SM-P610
* __Product:__ gta4xlwifieea
Crash log

```
io.reactivex.exceptions.OnErrorNotImplementedException: The exception was 
not handled due to missing onError handler in the subscribe() method 
call. Further reading: 
https://github.com/ReactiveX/RxJava/wiki/Error-Handling | 
android.database.sqlite.SQLiteConstraintException: UNIQUE constraint 
failed: bookmarks.path (code 2067 SQLITE_CONSTRAINT_UNIQUE[2067])
at 
io.reactivex.internal.observers.EmptyCompletableObserver.onError(EmptyCompletableObserver.java:50)
at 
io.reactivex.internal.operators.completable.CompletableSubscribeOn$SubscribeOnObserver.onError(CompletableSubscribeOn.java:74)
at 
io.reactivex.internal.operators.completable.CompletableFromCallable.subscribeActual(CompletableFromCallable.java:40)
at io.reactivex.Completable.subscribe(Completable.java:2309)
at 
io.reactivex.internal.operators.completable.CompletableSubscribeOn$SubscribeOnObserver.run(CompletableSubscribeOn.java:64)
at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
at 
io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
at 
io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at 
java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
at 
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
at 
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
at java.lang.Thread.run(Thread.java:919)
Caused by: android.database.sqlite.SQLiteConstraintException: UNIQUE 
constraint failed: bookmarks.path (code 2067 
SQLITE_CONSTRAINT_UNIQUE[2067])
at 
android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native 
Method)
at 
android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:1127)
at 
android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:790)
at 
android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:88)
at 
androidx.sqlite.db.framework.FrameworkSQLiteStatement.executeInsert(FrameworkSQLiteStatement.java:51)
at 
androidx.room.EntityInsertionAdapter.insert(EntityInsertionAdapter.java:64)
at 
com.amaze.filemanager.database.daos.BookmarkEntryDao_Impl$4.call(BookmarkEntryDao_Impl.java:97)
at 
com.amaze.filemanager.database.daos.BookmarkEntryDao_Impl$4.call(BookmarkEntryDao_Impl.java:92)
at 
io.reactivex.internal.operators.completable.CompletableFromCallable.subscribeActual(CompletableFromCallable.java:36)
... 10 more

```

----------------------------------------

### PR DESCRIPTION
**Pull request by TranceLove**
#### Issue tracker   
Fixes will automatically close the related issue

Fixes #2218

#### Release  
Addresses release/3.6
  
#### Test cases
- [ ] Covered
  
#### Manual testing
- [x] Done  
  
If yes,  
- Device: Fairphone 3
- OS: LineageOS 16.0(9.0)

#### Build tasks success  
Successfully running following tasks on local 
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

#### Additional Info
Database update required.

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/database/daos/BookmarkEntryDao.java [hunk 1]
```diff
@@ -47,7 +48,7 @@
 @Dao
 public interface BookmarkEntryDao {
 
-  @Insert
+  @Insert(onConflict = OnConflictStrategy.IGNORE)
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/daos/BookmarkEntryDao.java:51]: Will this ignore every insertion after the first one?

File: app/src/main/java/com/amaze/filemanager/database/UtilitiesDatabase.java [hunk 2]
```diff
@@ -328,6 +328,21 @@ public void migrate(@NonNull SupportSQLiteDatabase database) {
         }
       };
 
+  private static final Migration MIGRATION_4_5 =
+      new Migration(4, 5) {
+        @Override
+        public void migrate(@NonNull SupportSQLiteDatabase database) {
+          database.execSQL(
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/database/UtilitiesDatabase.java:335]: What happens to the old table, that could have indexes repeated

File: app/src/main/java/com/amaze/filemanager/database/daos/BookmarkEntryDao.java (hunk 1 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/database/daos/BookmarkEntryDao.java:51]: Would it be sensible if the same record having the same name and path get overwritten again and again, but it won't change the position of the bookmark at the drawer itself?

File: app/src/main/java/com/amaze/filemanager/database/UtilitiesDatabase.java (hunk 2 above)
Comment (TranceLove) [reply app/src/main/java/com/amaze/filemanager/database/UtilitiesDatabase.java:335]: And hell, I was even indexing the wrong table :(

File: app/src/main/java/com/amaze/filemanager/database/models/utilities/Bookmark.java [hunk 3]
```diff
@@ -23,13 +23,21 @@
...
  *
  * @see UtilitiesDatabase
  */
-@Entity(tableName = UtilitiesDatabase.TABLE_BOOKMARKS)
+@Entity(
+    tableName = UtilitiesDatabase.TABLE_BOOKMARKS,
+    indices = {
+      @Index(
+          name = "bookmarks_idx",
+          value = {"name", "path"},
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/database/models/utilities/Bookmark.java:38]: Can use `UtilsDatabase.COLUMN_NAME` and `UtilsDatabase.COLUMN_PATH`?

File: app/src/main/java/com/amaze/filemanager/database/models/utilities/Bookmark.java [hunk 4]
```diff
@@ -23,13 +23,21 @@
...
  * {@link Entity} representation of <code>bookmark</code> table in utilities.db.
  *
  * @see UtilitiesDatabase
  */
-@Entity(tableName = UtilitiesDatabase.TABLE_BOOKMARKS)
+@Entity(
+    tableName = UtilitiesDatabase.TABLE_BOOKMARKS,
+    indices = {
+      @Index(
+          name = "bookmarks_idx",
```
Comment (VishalNehra) [inline app/src/main/java/com/amaze/filemanager/database/models/utilities/Bookmark.java:37]: Can define a reference variable?

**Review by VishalNehra — CHANGES_REQUESTED** [review]

**Review by VishalNehra — APPROVED** [review]

**Pull request merged**